File: src/dotnet/Harness.Engine/EnvelopeValidation.cs
using System.Text.RegularExpressions;

namespace Harness.Engine;

/// <summary>Result of a contextual validation: ok, or the reason for rejection (for the corrective error).</summary>
public readonly record struct ValidationResult(bool Ok, string Reason)
{
    public static ValidationResult Pass { get; } = new(true, string.Empty);
    public static ValidationResult Fail(string reason) => new(false, reason);
}

/// <summary>
/// Deterministic, cheap predicates to validate whether the value the driver returned
/// meets the task's expectation — BEFORE persisting it and continuing the flow. Failed →
/// <see cref="TaskRegistry"/> returns a typed corrective error and the driver resends
/// (corrective loop, not silent termination).
///
/// Deep semantic validation is still the LLM judge's job during evaluation; only what's
/// checkable in code, at zero token cost, lives here.
/// </summary>
public static class EnvelopeValidation
{
    /// <summary>The first arg exists and is not empty/whitespace.</summary>
    public static Func<Envelope, ValidationResult> NotEmpty(string expectation) =>
        envelope => FirstArg(envelope) is { Length: > 0 }
            ? ValidationResult.Pass
            : ValidationResult.Fail($"The expected argument came back empty. Expected: {expectation}.");

    /// <summary>The first arg has at least <paramref name="count"/> non-empty lines (counting literal <c>\n</c>).</summary>
    public static Func<Envelope, ValidationResult> MinLines(int count, string expectation) =>
        envelope =>
        {
            var lines = Lines(FirstArg(envelope));
            return lines >= count
                ? ValidationResult.Pass
                : ValidationResult.Fail(
                    $"The argument has {lines} non-empty line(s), but the task expects at least {count}. Expected: {expectation}.");
        };

    /// <summary>The first arg contains at least one number.</summary>
    public static Func<Envelope, ValidationResult> ContainsNumber(string expectation) =>
        envelope => Regex.IsMatch(FirstArg(envelope), @"\d")
            ? ValidationResult.Pass
            : ValidationResult.Fail($"The argument does not contain any number. Expected: {expectation}.");

    /// <summary>The first arg matches the pattern (case-insensitive).</summary>
    public static Func<Envelope, ValidationResult> Matches(string pattern, string expectation) =>
        envelope => Regex.IsMatch(FirstArg(envelope), pattern, RegexOptions.IgnoreCase)
            ? ValidationResult.Pass
            : ValidationResult.Fail($"The argument does not match the expected format. Expected: {expectation}.");

    /// <summary>Composition: every predicate must pass; the first one that fails supplies the reason.</summary>
    public static Func<Envelope, ValidationResult> All(params Func<Envelope, ValidationResult>[] validators) =>
        envelope =>
        {
            foreach (var validator in validators)
            {
                var result = validator(envelope);
                if (!result.Ok)
                    return result;
            }

            return ValidationResult.Pass;
        };

    private static string FirstArg(Envelope envelope) =>
        envelope.Args is { Length: > 0 } args ? args[0].Trim() : string.Empty;

    // Artifacts travel as a single-line JSON string with literal \n (see the flows'
    // "Compact" notice) — counts both real and escaped line breaks.
    private static int Lines(string value) =>
        value.Split(["\n", "\\n"], StringSplitOptions.RemoveEmptyEntries)
            .Count(line => !string.IsNullOrWhiteSpace(line));
}

File: src/go/harnessengine/envelope_validation.go
package harnessengine

import (
	"fmt"
	"regexp"
	"strings"
)

// ValidationResult is the result of a contextual validation: ok, or the rejection reason
// (for the corrective error).
type ValidationResult struct {
	Ok     bool
	Reason string
}

// ValidationPass is the successful ValidationResult.
var ValidationPass = ValidationResult{Ok: true}

// ValidationFail builds a failing ValidationResult with the given reason.
func ValidationFail(reason string) ValidationResult {
	return ValidationResult{Ok: false, Reason: reason}
}

// Validator is a deterministic, cheap predicate over an Envelope's returned value —
// checked BEFORE persisting it and advancing the flow. Failed → TaskRegistry returns a
// typed corrective error and the driver resends (corrective loop, not silent termination).
//
// Deep semantic validation is still the LLM judge's job during evaluation; only what is
// checkable in code, at zero token cost, lives here.
type Validator func(Envelope) ValidationResult

// NotEmpty passes when the first arg exists and is not empty/whitespace.
func NotEmpty(expectation string) Validator {
	return func(e Envelope) ValidationResult {
		if firstArg(e) != "" {
			return ValidationPass
		}
		return ValidationFail(fmt.Sprintf("The expected argument came back empty. Expected: %s.", expectation))
	}
}

// MinLines passes when the first arg has at least count non-empty lines (counting literal \n too).
func MinLines(count int, expectation string) Validator {
	return func(e Envelope) ValidationResult {
		lines := countLines(firstArg(e))
		if lines >= count {
			return ValidationPass
		}
		return ValidationFail(fmt.Sprintf(
			"The argument has %d non-empty line(s), but the task expects at least %d. Expected: %s.", lines, count, expectation))
	}
}

var digitRegexp = regexp.MustCompile(`\d`)

// ContainsNumber passes when the first arg contains at least one digit.
func ContainsNumber(expectation string) Validator {
	return func(e Envelope) ValidationResult {
		if digitRegexp.MatchString(firstArg(e)) {
			return ValidationPass
		}
		return ValidationFail(fmt.Sprintf("The argument does not contain any number. Expected: %s.", expectation))
	}
}

// Matches passes when the first arg matches pattern (case-insensitive).
func Matches(pattern, expectation string) Validator {
	re := regexp.MustCompile("(?i)" + pattern)
	return func(e Envelope) ValidationResult {
		if re.MatchString(firstArg(e)) {
			return ValidationPass
		}
		return ValidationFail(fmt.Sprintf("The argument does not match the expected format. Expected: %s.", expectation))
	}
}

// AllOf composes validators: all must pass; the first failure supplies the reason.
func AllOf(validators ...Validator) Validator {
	return func(e Envelope) ValidationResult {
		for _, v := range validators {
			if result := v(e); !result.Ok {
				return result
			}
		}
		return ValidationPass
	}
}

func firstArg(e Envelope) string {
	if len(e.Args) == 0 {
		return ""
	}
	return strings.TrimSpace(e.Args[0])
}

// countLines counts non-empty lines, splitting on both literal "\n" markers and real
// newlines — artifacts travel as a single-line JSON string with literal \n (see the
// "Compact" warning in the flows).
func countLines(value string) int {
	normalized := strings.ReplaceAll(value, "\\n", "\n")
	count := 0
	for _, line := range strings.Split(normalized, "\n") {
		if strings.TrimSpace(line) != "" {
			count++
		}
	}
	return count
}
